distributed/utils.py

```diff
@@ -1608,3 +1608,23 @@ def recursive_to_dict(
         return repr(obj)
     finally:
         tok.var.reset(tok)
+
+
+T = TypeVar("T")
+
+
+async def ensure_cancellation(coro: CoroutineType[None, None, T]) -> T:
+    """Ensure that the wrapped coro will raise a CancelledError even if its result is already set"""
+    watcher = asyncio.Event()
+
+    task = asyncio.create_task(coro)
+    task.add_done_callback(lambda _: watcher.set())
+
+    try:
+        await watcher.wait()
+    except asyncio.CancelledError:
+        task.cancel()
+        await watcher.wait()
+        raise
```

Comment:
I'm not entirely sure if I need to await the task itself here. There are no warnings 

cc @graingert